### ISSUE TITLE
Compilation failure on 0.3.3rc1 using Metal

### ISSUE THREAD

**Issue by KuzMenachem** [plaidml/plaidml#128]
When using this code I get the following error
I'm using a Mac with an AMD R9 M395 on 10.13.4

Thanks for your help

```
import numpy as np
import plaidml.keras
plaidml.keras.install_backend()
from keras.models import Model
from keras.layers import Dense, Input
from keras.optimizers import SGD

x = np.random.random((10, 1))
y = np.round(np.random.random((10, 1)))

input = Input(shape=(1,))

output = Dense(1, activation='sigmoid')(input)

model = Model(inputs=[input], outputs=[output])

sgd = SGD(lr=1)

model.compile(loss='binary_crossentropy',
              optimizer=sgd,
              metrics=['accuracy']
              )

history = model.fit(x, y, epochs=1000)
```

```INFO:plaidml:Opening device "metal_amd_radeon_r9_m395.0"
Epoch 1/1000
ERROR:plaidml:Compiler::Build> Compilation failure:
Compilation failed: 

program_source:23:5: error: type 'device ptrdiff_t *' (aka 'device long *') is not valid for attribute 'buffer'
    device ptrdiff_t* X_T1 [[ buffer(0) ]],
    ^                         ~~~~~~~~~
program_source:23:12: note: type 'ptrdiff_t' (aka 'long') cannot be used in buffer pointee type
    device ptrdiff_t* X_T1 [[ buffer(0) ]],
           ^
program_source:24:5: error: type 'const device ptrdiff_t *' (aka 'const device long *') is not valid for attribute 'buffer'
    device const ptrdiff_t* X_I_0 [[ buffer(1) ]],
    ^                                ~~~~~~~~~
program_source:24:18: note: type 'ptrdiff_t' (aka 'long') cannot be used in buffer pointee type
    device const ptrdiff_t* X_I_0 [[ buffer(1) ]],
                 ^
program_source:30:7: warning: unused variable 'tid'
  int tid = _tid;
      ^
program_source:70:7: warning: unused variable 'tid'
  int tid = _tid;
      ^

Traceback (most recent call last):
  File "test.py", line 26, in <module>
    history = model.fit(x, y, epochs=1000, verbose=1)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/keras/engine/training.py", line 1705, in fit
    validation_steps=validation_steps)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/keras/engine/training.py", line 1236, in _fit_loop
    outs = f(ins_batch)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plaidml/keras/backend.py", line 151, in __call__
    self._invoker.invoke()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plaidml/__init__.py", line 1426, in invoke
    return Invocation(self._ctx, self)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plaidml/__init__.py", line 1432, in __init__
    self._as_parameter_ = _lib().plaidml_schedule_invocation(ctx, invoker)
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plaidml/__init__.py", line 716, in _check_err
    self.raise_last_status()
  File "/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/plaidml/library.py", line 131, in raise_last_status
    raise self.last_status()
plaidml.exceptions.Internal: The exception chain appears to be corrupt
```

**Comment by flaub** [comment]
This indicates that the network you are trying to run uses a 64bit integer. This isn’t supported yet with out Metal backend. If possible try using 32bit integers. I’m less familiar with numpy but perhaps the random function can be given a precision?

**Comment by flaub** [comment]
To clarify, Metal doesn’t support 64bit input and output buffers. However 64bit integers can be used as a data type for interior variables. 

Also it’s possible to fix our metal backend to support 64 bit buffers by using a workaround (on our end, by changing our code emitter).

**Comment by KuzMenachem** [comment]
I just checked with a couple different data types, including float32 and (the default) float64, int32 and int64 and bool. The error still persists. I also tried running plaidvision with the mobilenet/webcam demo, which gives a very similar error.

**Comment by flaub** [comment]
TL;DR: an upcoming release should fix this issue.

I did some digging and found out that training with keras always uses an int64 to track the number of iterations that have run. In the next release, 64-bit data types can now be used for input/output buffers.

However, getting training to work uncovered a pretty serious memory leak. So the upcoming release also plugs these leaks.

I'll close this issue as soon as the next release with these fixes is out and available.

**Comment by KuzMenachem** [comment]
Thanks!

**Comment by flaub** [comment]
We've released a new version that hopefully fixes this issue. Could you try version 0.3.3rc3?

**Comment by KuzMenachem** [comment]
It's fixed 👍

**Comment by flaub** [comment]
Great!
